### ISSUE TITLE
Pantheon is signing blocks with nonce 0xfff.. in Clique chains by default, casting votes for 0x00.. as a validator 

### ISSUE THREAD

**Issue by JohnnySheffield** [PegaSysEng/pantheon#762]
### Description
When Pantheon node is an authority in a clique-based network, it is defaulting to `0xfff...` for nonce value, effectively casting votes for `0x0000....`

From the clique spec:

1.) Nonce value must be `NONCE_DROP` (`0x0000000000000000`) when voting to remove a signer,
2.) Nonce value must be `NONCE_AUTH` (`0xffffffffffffffff`) when voting to add a signer
3.) Nonce value must be `NONCE_DROP` on checkpoint block.
4.) All other times must not take up any other value except `NONCE_DROP` or `NONCE_AUTH`.

This can be seen in Goerli chain:

```
> clique.getSnapshot()
{
  hash: "0x59f7d6d47fbd58aba6038af98678014fc6f89cf842af1f44d0a0b21cdc42bce0",
  number: 22584,
  recents: {
    22581: "0xe0a2bd4258d2768837baa26a28fe71dc079f84c7",
    22582: "0x000000568b9b5a365eaa767d42e74ed88915c204",
    22583: "0x22ea9f6b28db76a7162054c05ed812deb2f519cd",
    22584: "0x4c2ae482593505f0163cdefc073e81c63cda4107"
  },
  signers: {
    0x000000568b9b5a365eaa767d42e74ed88915c204: {},
    0x22ea9f6b28db76a7162054c05ed812deb2f519cd: {},
    0x4c2ae482593505f0163cdefc073e81c63cda4107: {},
    0x8b24eb4e6aae906058242d83e51fb077370c4720: {},
    0xa8e8f14732658e4b51e8711931053a8a69baf2b1: {},
    0xd9a5179f091d85051d3c982785efd1455cec8699: {},
    0xe0a2bd4258d2768837baa26a28fe71dc079f84c7: {}
  },
  tally: {
    0x0000000000000000000000000000000000000000: {
      authorize: true,
      votes: 1
    }
  },
  votes: [{
      address: "0x0000000000000000000000000000000000000000",
      authorize: true,
      block: 22583,
      signer: "0x22ea9f6b28db76a7162054c05ed812deb2f519cd"
  }]
}
```
### Acceptance Criteria
* When block is not casting a vote, nonce value should be `NONCE_DROP`

**Expected behavior:** 

* default nonce value should be `NONCE_DROP`

**Actual behavior:**
* default nonce value is `NONCE_AUTH`

### Additional Information
Thanks to @karalabe for providing feedback on the expected default nonce value and getting clique.getSnapshot() logs

**Comment by karalabe** [comment]
Just to add to this, the spec doesn't really handle `NONCE_DROP/0x00` differently either from the perspective of voting (i.e. it still counts as a "drop the `0x0000000000000000000000000000000000000000` signer", but since that's not a signer really, the vote is discarded without being interpreted).

**Comment by rain-on** [comment]
From testing, it appears the Pantheon client performs (largely) in line specification (i.e. default to 0x0 nonce and 0x0 coinbase) - however:
The Pantheon Validator on Goerli had  the following outstanding Proposals (injected via JSON):
`{
  "jsonrpc" : "2.0",
  "id" : 1,
  "result" : {
    "0x22ea9f6b28db76a7162054c05ed812deb2f519cd" : true,
    "0xa8e8f14732658e4b51e8711931053a8a69baf2b1" : true,
    "0x000000568b9b5a365eaa767d42e74ed88915c204" : true,
    "0x4c2ae482593505f0163cdefc073e81c63cda4107" : true,
    "0xd9a5179f091d85051d3c982785efd1455cec8699" : true,
    **"0x0000000000000000000000000000000000000000" : true,**
    "0x8b24eb4e6aae906058242d83e51fb077370c4720" : true
  }
}`

Thus, the client was attempting to have 0x0 authorised (and would continue to do this until it was accepted).

The spurious proposal has now been removed, and nonces from 24599 are now 0x0.

Having said that - it was identified that Pantheon doesn't ensure a null-vote on Epoch blocks.

**Comment by rain-on** [comment]
To follow up on the 0x0 beneficiary - the Clique EIP states:
`beneficiary: Address to propose modifying the list of authorized signers with.
Should be filled with zeroes normally, modified only while voting.`

From this, it was determined Pantheon would discard the vote from any block containing a beneficiary of 0x0 (i.e. the nonce value is immaterial).

From @karalabe comment above, it appears Geth actions the vote regardless of the coinbase's value (thus the Auth-0x0 appearing in the snapshot).

There are a few different ways forward to overcome the discrepancy in behaviour:
1. Modify spec to clarify the voting process (i.e. inline with per Geth's behaviour) - i.e. every block construes a vote - so the default is to "Drop 0x0"
1a. Modify Pantheon's behaviour to follow suit.
2. Leave spec as is, modify Geth to discard votes with a 0x0 beneficiary.

**Comment by JohnnySheffield** [comment]
@rain-on it seems geth is not counting votes if the nonce value is `NONCE_DROP` and the `beneficiary` is `0x000...`:

```
> clique.getSnapshot()
{
  hash: "0x59ffc3063c6f5cd9040523a7de6151a1cbd9445332ef296fb5bb2cb500c3f5fd",
  number: 26983,
  recents: {
    26980: "0xe0a2bd4258d2768837baa26a28fe71dc079f84c7",
    26981: "0x8b24eb4e6aae906058242d83e51fb077370c4720",
    26982: "0xd9a5179f091d85051d3c982785efd1455cec8699",
    26983: "0x000000568b9b5a365eaa767d42e74ed88915c204"
  },
  signers: {
    0x000000568b9b5a365eaa767d42e74ed88915c204: {},
    0x22ea9f6b28db76a7162054c05ed812deb2f519cd: {},
    0x4c2ae482593505f0163cdefc073e81c63cda4107: {},
    0x8b24eb4e6aae906058242d83e51fb077370c4720: {},
    0xa8e8f14732658e4b51e8711931053a8a69baf2b1: {},
    0xd9a5179f091d85051d3c982785efd1455cec8699: {},
    0xe0a2bd4258d2768837baa26a28fe71dc079f84c7: {}
  },
  tally: {},
  votes: []
}
```

**Comment by karalabe** [comment]
@JohnnySheffield Well, the account is not an authorized signer, so the vote is invalid and is discarded. Same way we won't count votes for `NONCE_AUTH` if the address is already a signer. Keeps the voting state a lot cleaner.

Not sure if this is in the spec, if not, I think we should add it as it's the correct behavior imho. Otherwise you get strange side effects when the same signer gets added/removed multiple times in a single epoch.

**Comment by rain-on** [comment]
@karalabe So it sounds like Pantheon is doing something a little bit wrong, in that Address 0x00 should not be treated as a special case - i.e. every block contains a vote, by default that vote is 'DROP 0x00' (which should not get actioned as 0x00 should not be a validator).

This implies that if a vote is not valid at block reception (eg. DROP a non-validator) it should be discarded, and not buffered until epoch (which is what Pantheon is currently doing).

I am assuming that if a node votes in the opposite direction to their prior vote (and before quorum is reached), the preceding vote should be removed?

**Comment by jframe** [comment]
Pantheon has been updated so that will now discard proposals for address 0 (https://github.com/PegaSysEng/pantheon/pull/831). This is consistent with how geth is handling proposals for 0x0.

We have also fixed a bug so that on the epoch block we are using a beneficiary filled with zeros (https://github.com/PegaSysEng/pantheon/pull/833).

Both of these changes are included in latest 1.0.0 release and also in master.
